packages/query-store/src/QueryStore.ts

```diff
@@ -0,0 +1,340 @@
+import { AbstractStore } from './abstract.store.js';
+import type {
+	AggFunction,
+	ColumnMetadata,
+	QueryResult,
+	QueryStoreOpts,
+	QueryStoreValue,
+	Runner
+} from './types.js';
+
+import { Query, sql, count } from '@uwdata/mosaic-sql';
+import { buildId } from './utils/buildId.js';
+import { handleMaybePromise } from './utils/handleMaybePromise.js';
+import { mutations } from './mutations/index.js';
+
+export class QueryStore extends AbstractStore<QueryStoreValue> {
+	/** Indicate that QueryStore is readable like an array */
+	[index: number]: QueryResult;
+
+	/** Internal Query Builder */
+	readonly #query = new Query();
+
+	/** Currently Held Values */
+	#values: QueryResult[] = [];
+
+	value = () => this.#proxied;
+
+	/** Query Execution Function */
+	readonly #exec: Runner;
+
+	/**
+	 * A Proxy wrapper around the QueryStore instance.
+	 * It is used to intercept access to numeric indices and the 'length' property.
+	 * This Proxy is responsible for triggering data fetching.
+	 * When any numeric index or the 'length' property is accessed, the Proxy command triggers
+	 * either the #update() or the #updateLength() function to asynchronously load or update data.
+	 * Hence, only through this Proxy (i.e., #proxied), can the data fetching and length update
+	 * process be triggered. Accessing the QueryStore directly does not trigger these functionalities,
+	 * and hence is not recommended for normal use.
+	 */
+	readonly #proxied: QueryStoreValue;
+	get proxy() {
+		return this.#proxied;
+	}
+
+	/** Text of the query represented by this store */
+	get text() {
+		// TODO: This needs a formatter
+		return this.#query.toString();
+	}
+
+	/** Name and Type information about the result columns */
+	#columns: ColumnMetadata[] = [];
+	get columns() {
+		return Array.from(this.#columns);
```

Comment:
Yep; prevents the actual array from being modified. This may need to be a little more resilient because the objects themselves may be mutable. Could just freeze them though